### ISSUE TITLE
`partial` incorrectly narrows return type when typevars are used in function-wrapper signatures

### ISSUE THREAD

**Issue by TTsangSC** [astral-sh/ty#3467]
### Summary

I'm struggling to further simplify the MWE and the error looks a bit different from what I get on my real code on local... but this may be interesting in its own right. Essentially, `ty` correctly narrows the type of the `partial` when the supplied function wrapper uses `Any` in its signature, but not when it uses a `TypeVar`.

Example
----
(Playground: https://play.ty.dev/775c0813-217b-4c3e-ba8a-4d2e01965cc3)

```python
from collections.abc import Callable
from functools import partial
from typing import Any, ParamSpec, TypeVar, reveal_type


T = TypeVar('T')
PS = ParamSpec('PS')


def process_data_typed(
    x: str, get_maybe_data: Callable[PS, tuple[T, ...] | None], /,
    *args: PS.args, **kwargs: PS.kwargs
) -> tuple[T, ...] | None:
    raw = get_maybe_data(*args, **kwargs)
    if raw is None:
        return None
    return raw[:-1]


def process_data_anys(
    x: str, get_maybe_data: Callable[PS, tuple[Any, ...] | None], /,
    *args: PS.args, **kwargs: PS.kwargs
) -> tuple[Any, ...] | None:
    raw = get_maybe_data(*args, **kwargs)
    if raw is None:
        return None
    return raw[:-1]


def do_something_typed(get_maybe_data: Callable[[], tuple[T, ...] | None]) -> None:
    get_wrapper: Callable[[], tuple[T, ...] | None] = reveal_type(partial(process_data_typed, 'foo', get_maybe_data))


def do_something_anys(get_maybe_data: Callable[[], tuple[Any, ...] | None]) -> None:
    get_wrapper: Callable[[], tuple[Any, ...] | None] = reveal_type(partial(process_data_anys, 'foo', get_maybe_data))
```

Expected
----
```
Revealed type: `partial[() -> tuple[T, ...] | None]`
Revealed type: `partial[() -> tuple[Any, ...] | None]`
```

Actual
----
```
Revealed type: `partial[None]` (revealed-type) [Ln 31, Col 67]
Argument to class `partial` is incorrect: Expected `(...) -> None`, found `def process_data_typed[**PS, T](x: str, get_maybe_data: (**PS) -> tuple[T, ...] | None, /, *args: PS.args, **kwargs: PS.kwargs) -> tuple[T, ...] | None` (invalid-argument-type) [Ln 31, Col 75]
Revealed type: `partial[() -> tuple[Any, ...] | None]` (revealed-type) [Ln 35, Col 69]
```
Note how:
- Both `process_data_typed()` and `process_data_anys()` pass type-checking on their own.
- While `Any` results in the correct narrowing (to the extent that is possible) of the return type of the `partial`, when the `TypeVar` `T` is used, `ty` incorrectly asserts that the `partial` should return `None` and thus complains about the typing on the argument ~~`get_maybe_data`~~ **EDIT: `process_data_typed`**.

Note
----
I do understand that `partial` is not fully supported as-is (#1536), but it is nonetheless a surprise when the behaviors diverge in such a way.

### Version

1b96fc10e (as seen on Playground)

**Comment by carljm** [comment]
ty behaves as expected on the `Any` version, but not the generic version. I suspect that this is because the generics support in partial hasn't landed yet - @charliermarsh is that right?

Simplified version showing just the failing example, without extraneous arguments:

```py
from collections.abc import Callable
from functools import partial
from typing import ParamSpec, TypeVar, reveal_type

T = TypeVar('T')
PS = ParamSpec('PS')

def process_data_typed(
    get_maybe_data: Callable[PS, tuple[T, ...] | None], /,
) -> tuple[T, ...] | None:
    raise NotImplementedError()

def do_something_typed(get_maybe_data: Callable[[], tuple[T, ...] | None]) -> None:
    get_wrapper: Callable[[], tuple[T, ...] | None] = reveal_type(partial(process_data_typed, get_maybe_data))
```

**Comment by charliermarsh** [comment]
It doesn't seem to be solved by any of the open PRs, though haven't dug into why.

**Comment by charliermarsh** [comment]
Codex suggests that this isn't specific to partial, since this also errors:

```python
from collections.abc import Callable
from typing import reveal_type

def generic[T](value: T) -> tuple[T, ...] | None:
    raise NotImplementedError

def return_type_of[R](function: Callable[..., R]) -> R:
    raise NotImplementedError

# Emits:
# Expected (...) -> None,
# found def generic[T](value: T) -> tuple[T, ...] | None
reveal_type(return_type_of(generic))
```
